radicle/src/ssh.rs

```diff
@@ -184,6 +252,165 @@ impl radicle_ssh::key::Private for SecretKey {
+        let sig_version = r.read_u32()?;
+        if sig_version > 1 {
+            return Err(ExtendedSignatureError::UnsupportedVersion(sig_version));
+        }
+
+        Ok(ExtendedSignature {
+            sig_version,
+            publickey: PublicKey::read_ssh(r)
+                .map_err(|_| ExtendedSignatureError::PublicKeyEncoding())?,
+            namespace: r.read_string()?.into(),
+            reserved: r.read_string()?.into(),
+            hash_algorithm: r.read_string()?.into(),
+            signature: crypto::Signature::read_ssh(r)
+                .map_err(|_| ExtendedSignatureError::PublicKeyEncoding())?,
+        })
+    }
+
+    fn write_ssh<E: Encoding>(&self, buf: &mut E) {
+        buf.extend_u32(self.sig_version);
+        let _ = &self.publickey.write_ssh(buf);
+        buf.extend_ssh_string(&self.namespace);
+        buf.extend_ssh_string(&self.reserved);
+        buf.extend_ssh_string(&self.hash_algorithm);
+        let _ = &self.signature.write_ssh(buf);
+    }
+}
+
+fn copy_without<T>(arg: &[T], not_include: T) -> Vec<T>
+where
+    T: Copy + PartialEq,
+{
+    let mut buf = Vec::<T>::new();
+    for c in arg {
+        if c != &not_include {
+            buf.push(*c);
+        }
+    }
+    buf
+}
+
+impl ExtendedSignature {
+    const ARMORED_HEADER: &[u8; 29] = b"-----BEGIN SSH SIGNATURE-----";
+    const ARMORED_FOOTER: &[u8; 27] = b"-----END SSH SIGNATURE-----";
+    const ARMORED_WIDTH: usize = 70;
+    const MAGIC_PREAMBLE: &[u8; 6] = b"SSHSIG";
+
+    pub fn from_armored(s: &[u8]) -> Result<Self, ExtendedSignatureError> {
+        let s = s
+            .strip_prefix(Self::ARMORED_HEADER)
+            .ok_or(ExtendedSignatureError::MissingHeader())?;
+        let s = s
+            .strip_suffix(Self::ARMORED_FOOTER)
+            .ok_or(ExtendedSignatureError::MissingFooter())?;
+        let s = copy_without(s, b'\n');
+
+        let buf = base64::decode(s)?;
+        let mut reader = buf.reader(0);
+
+        let preamble: [u8; 6] = reader.read_bytes()?;
+        if &preamble != Self::MAGIC_PREAMBLE {
+            return Err(ExtendedSignatureError::MagicPreamble(preamble));
+        }
+
+        let sig = ExtendedSignature::read_ssh(&mut reader)?;
+        Ok(sig)
+    }
+
+    pub fn to_armored(&self) -> Vec<u8> {
+        let mut v: encoding::Buffer = Vec::<u8>::new().into();
+
+        v.extend(Self::MAGIC_PREAMBLE);
+        self.write_ssh(&mut v);
+        let mut body: Vec<u8> = base64::encode(v).into();
+
+        let mut armored = Self::ARMORED_HEADER.to_vec();
+        armored.push(b'\n');
+        while !body.is_empty() {
+            let at = if body.len() >= Self::ARMORED_WIDTH {
+                Self::ARMORED_WIDTH
+            } else {
+                body.len()
+            };
+            let (l, rest) = body.split_at(at);
+
+            let l: Vec<u8> = l.into();
+            armored.extend(l);
+            armored.push(b'\n');
+
+            body = rest.to_vec();
+        }
```

Comment:
This loop can be replaced with something like `body.iter().chunks(Self::ARMORED_WIDTH).collect().join(b'\n')`